homeassistant/components/tts/__init__.py

```diff
@@ -0,0 +1,143 @@
+"""
+Support for text to speech in Home Assistant.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/tts/
+"""
+
+import logging
+import re
+import hashlib
+import os.path
+from functools import partial
+
+from homeassistant import core
+import homeassistant.bootstrap as bootstrap
+from homeassistant.helpers import config_per_platform
+from homeassistant.const import (ATTR_ENTITY_ID, SERVICE_PLAY_MEDIA, CONF_NAME)
+from homeassistant.components.media_player import MEDIA_TYPE_MUSIC
+from homeassistant.config import load_yaml_config_file
+
+DEPENDENCIES = ['http']
+
+DOMAIN = "tts"
+
+SERVICE_TTS = "tts"
+
+ATTR_TEXT = "text"
+ATTR_ALLOW_CACHE = "allow_cached_file"
+ATTR_PLAY = "play"
+ATTR_LANGUAGE = "language"
+ATTR_RATE = "rate"
+ATTR_CODEC = "codec"
+ATTR_FORMAT = "format"
+ATTR_MEDIA_CONTENT_ID = "media_content_id"
+ATTR_MEDIA_CONTENT_TYPE = "media_content_type"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup(hass, config):
+    """Setup the tts service."""
+    success = False
+
+    descriptions = load_yaml_config_file(
+        os.path.join(os.path.dirname(__file__), 'services.yaml'))
+
+    tts_dir = hass.config.path(DOMAIN)
+
+    if not os.path.exists(tts_dir):
+        os.makedirs(tts_dir)
+
+    for platform, p_config in config_per_platform(config, DOMAIN, _LOGGER):
+        tts_implementation = bootstrap.prepare_setup_platform(
+            hass, config, DOMAIN, platform)
+
+        if tts_implementation is None:
+            _LOGGER.error("Unknown tts service specified.")
+            continue
+
+        tts_engine = tts_implementation.get_engine(hass, p_config)
+
+        if tts_engine is None:
+            _LOGGER.error("Failed to initialize tts service %s",
+                          platform)
+            continue
+
+        def generate_speech(tts_service, platform_name, call):
+            """Handle sending tts message service calls."""
+            text = call.data.get(ATTR_TEXT)
+            entity_id = call.data.get(ATTR_ENTITY_ID)
+            allow_cached_file = call.data.get(ATTR_ALLOW_CACHE, True)
+            play_file = call.data.get(ATTR_PLAY, True)
+            language = call.data.get(ATTR_LANGUAGE, config.get(ATTR_LANGUAGE,
```

Comment:
Instead of providing default values, we should have the platforms figure out default values.